### ISSUE TITLE
Include() causing 'Sequence contains no matching element' EF Core 3.0 preview 7 and 8

### ISSUE THREAD

**Issue by dotnetshadow** [dotnet/efcore#17137]
Describe what is not working as expected.
In EF Core 3.0 preview 5 query was working, but in EF Core 3.0 preview 7 and 8 it fails with no sequence found

Been working with an owner of the open source project *KalikoCMS*
Recently while upgrading to .net core 3.0 with EF core 3.0 noticed there has been a change in one of the queries
With version Ef core 3.0 preview5 things are ok, but with any preview after that there is an exception.

```
_contentTypeRepository.GetAll().Include(x => x.Properties).ToList();

public virtual IQueryable<TEntity> GetAll() {
            return _context.Set<TSource>().Select(Map()).AsNoTracking();
}

public override Expression<Func<LegacyPageTypeEntity, ContentTypeEntity>> Map() {
            return x => new ContentTypeEntity {
                ContentTypeId = ToGuid(x.PageTypeId),
                DefaultChildSortDirection = x.DefaultChildSortDirection,
                ContentProviderId = Guid.Empty,
                DefaultChildSortOrder = x.DefaultChildSortOrder,
                DisplayName = x.DisplayName,
                Name = x.Name,
                ShowInAdmin = true
            };
        }
```
It seems to get the with GetAll(), but it doesn't like the Include(), any ideas if something may of changed between previews?

The project can be found here:
https://github.com/KalikoCMS/KalikoCMS.NextGen/blob/netcore3.0-experimental/KalikoCMS/Services/Initialization/ContentTypeSynchronizer.cs#L33

### Exception message:
Sequence contains no matching element
```
Stack trace:
exception = {"Sequence contains no matching element"}
"   at System.Linq.ThrowHelper.ThrowNoMatchException()\r\n   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)\r\n   at Microsoft.EntityFrameworkCore.Query.ReplacingExpressionVisitor.VisitMember(MemberExpression memberExpression)\r\n   at System.Linq.Expressions.MemberExpression.Accept(ExpressionVisitor visitor)\r\n   at Microsoft.EntityFrameworkCore.Query.ReplacingExpressionVisitor.Visit(Expression expression)\r\n   at Microsoft.EntityFrameworkCore.Query.ReplacingExpressionVisitor.Replace(Expression original, Expression replacement, Expression tree)\r\n   at Microsoft.EntityFrameworkCore.Query.NavigationExpansion.Internal.NavigationExpandingVisitor.ProcessInclude(MethodCallExpression methodCallExpression)\r\n   at Microsoft.EntityFrameworkCore.Query.NavigationExpansion.Internal.NavigationExpandingVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)\r\n   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)\r\n   at Microsoft.EntityFrameworkCore.Query.QueryOptimizer.Visit(Expression query)\r\n   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)\r\n   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator()\r\n   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.IncludableQueryable`2.GetEnumerator()\r\n   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)\r\n   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)\r\n   at KalikoCMS.Services.Initialization.ContentTypeSynchronizer.SynchronizeContentTypes() in D:\\Programming\\Projects\\Libraries\\KalikoCMS.NextGen\\KalikoCMS.NextGen\\KalikoCMS\\Services\\Initialization\\ContentTypeSynchronizer.cs:line 36\r\n   at KalikoCMS.Services.Content.ContentTypeResolver.Initialize() in D:\\Programming\\Projects\\Libraries\\KalikoCMS.NextGen\\KalikoCMS.NextGen\\KalikoCMS\\Services\\Content\\ContentTypeResolver.cs:line 36\r\n   at KalikoCMS.Services.Initialization.InitializationService.Initialize() in D:\\Programming\\Projects\\Libraries\\KalikoCMS.NextGen\\KalikoCMS.NextGen\\KalikoCMS\\Services\\Initialization\\InitializationService.cs:line 44"
```

### Steps to reproduce
1. Download the project
https://github.com/KalikoCMS/KalikoCMS.NextGen/tree/netcore3.0-experimental
[KalikoCMS.zip](https://github.com/aspnet/EntityFrameworkCore/files/3499333/KalikoCMS.zip) (**Database**)
2. Change the **EF core 3.0** reference from preview **5** to **8** in **KalioCMS.Legacy** and **KalikoCMS.Data.SqlServer** projects

**Line in question**
https://github.com/KalikoCMS/KalikoCMS.NextGen/blob/netcore3.0-experimental/KalikoCMS/Services/Initialization/ContentTypeSynchronizer.cs#L33

**Side Note:**
Also even if I wanted to reference the CMS  project which uses EF Core 3.0 preview5 in an asp.net core app that has EF Core 3.0 preview7 it seems it still causes issues.

### Further technical details
EF Core version: 3.0.0-preview8.19405.11
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: 
IDE: (e.g. Visual Studio 2019 16.3 preview 2)

**Comment by smitpatel** [comment]
Include after nominal type projection does not work. See #12737

The exception you are hitting. I am fixing it in my current work. It would end up throwing exception that, include specified on non-entity.

**Comment by dotnetshadow** [comment]
@smitpatel Cheers, so this is by design?

I'm not too good with EF Core, so trying to understand the issue a little better.

1. From my understanding going forward an exception will be thrown for these types of queries?
2. What steps can I take now to get this to work like it was working in preview5?
3. It seems the queries will need to be rewritten to have the include() before projections?

Appreciate all the help

**Comment by smitpatel** [comment]
1. Yes, an exception will be thrown.
2. For your case, I am not able to understand how would you do an include. Since you are constructing `ContentTypeEntity` yourself. For EF perspective it is an unknown type. The concept of Include (using navigation to load related data), does not make sense for arbitrary type.
3. Yes, query needs to be rewritten to have Include before you change the shape of query root. Though in your case, again, since you are constructing your own type, include would be just ignored.

**Comment by dotnetshadow** [comment]
Cheers,

I haven't really contributed to the code, just trying to help the owner get it ready for .net core 3.0

I guess the author is trying to project some properties and then include a navigation property which is a collection of another type. He is using generics, so was perhaps trying to make it easy to use an include once he had the projection.

It's interesting though how it's ever been working with EF Core 3.0 preview5?
Basically I downloaded the project changed reference to EF Core 3.0 preview8 and it stopped working. Unless this change was done between those previews?

**Comment by smitpatel** [comment]
In past, we used to guess entity type based on the ClrType. Which was pretty fragile and confusing with owned types where clrType does not correspond to single entity. So we made experience consistent in the sense that, once you project using custom projection, (i.e. no reference to the entity parameter in your projection), we will treat the type as opaque to us.
